### PR TITLE
Remove data protection keys from Azure blob storage when deleting tenant

### LINKED ISSUE #16799: Protection Key is not removed during tenant removal
### Describe the bug
When we remove a tenant using the UI, the database and the host container are cleared of the tenant. But the Protection key is not removed.

### Orchard Core version

Add the version of the Orchard Core NuGet packages you use, or the commit hash if you can reproduce this with the source code.

### To Reproduce
1. Create a tenant
2. Set it up
3. Disable it
4. Remove it
5. Check the protection key `DataProtectionKeys.xml` it will exists. I tested this when data protection are stored in Azure Blob storage "not sure if locally we have the same problem".

### Expected behavior
I would expect everything related to the tenant being removed to be removed.

### PR DESCRIPTION
**Pull request by microposmp**
Fixes #16799

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:33]:
```suggestion
    /// <param name="context">The <see cref="ShellRemovingContext"/>.</param>
```

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by Piedone** [comment]
When addressing review feedback, please adhere to the following:

- Please update your pull request according to feedback until it is approved by one of the core team members.
- Apply [suggested changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) directly so the reviewer doesn't have to eyeball the changes. These resolve themselves after applying them, and that's fine.
- Don't resolve other conversations so it's easier to track for the reviewer. Then, the reviewer will resolve them.
- Feel free to mark conversations that you addressed to keep track of them with an emoji or otherwise, just don't resolve them.
- Please keep conversations happening in line comments in those convos, otherwise, communication will be a mess. If you have trouble finding them, see [this video](https://github.com/OrchardCMS/OrchardCore/pull/14749#issuecomment-1917976028).
- When you're done addressing all feedback of a review, click ["Re-request review"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review) in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs]: Moved registration so it's only made when connection string is configured.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs]: Updated as suggested

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs]: Removed docs from constructor.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:37]: Yes, the tenant name is removed together with the blob.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs]: Removed returns from docs section.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: Why are you creating an instance of BlobClient here instead of using `IShellsFileStore`

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs:43]: Why singleton? I think transient is enough

Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs]: Totally agree. Transient is enough here.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs]: All other `IModularTenantEvents` implementations are registered as scoped though. That's the safest default in any case, and judging from how `ShellScope` uses these, also the appropriate one.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: `IShellsFileStore` is about storing shell settings (i.e. what's normally in `App_Data\Sites\Default\appsettings.json` and similar), so that's not really applicable here. But `BlobFileStore` could be used instead indeed.

**Review by Piedone — CHANGES_REQUESTED** [review]
See https://github.com/OrchardCMS/OrchardCore/pull/16839#discussion_r1790748698 and https://github.com/OrchardCMS/OrchardCore/pull/16839#discussion_r1790755391.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: yes any implementation of `IShellsFileStore` should be used instead. I think we should use the same instance used when creating the tenant.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: But `IShellsFileStore` is not for this as I elaborated above. It'll only have an instance if [Azure Shells Configuration Provider](https://docs.orchardcore.net/en/latest/reference/core/Shells/#azure-shells-configuration-provider) is used, but that feature is independent of [`OrchardCore.DataProtection.Azure`](https://docs.orchardcore.net/en/latest/reference/modules/DataProtection.Azure/) and connects to a different Azure Blob Storage container.

Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: But I think that will be a breaking change. The configuration used for the OrchardCore.DataProtection.Azure module and the for example the Media module that uses BlobFileStore differs. The file store configuration have a BasePath that I think is treated as a virtual folder when saving blobs. The DataProtection have a BlobName that points to a specific blob.

Here only a specific blob belonging to the deleted tenant should be deleted. It doesn't need the functionality from the more feature rich FileStore.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs
Comment (microposmp) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs]: Changed to scoped lifetime.

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: Looking at the implementation of `OrchardCore.DataProtection.Azure`, it's fine to use `BlobClient` as you are doing for now. 

At some point, we should refactor the implementation to use `IDataProtectionFileStore` this would allow us to reuse the code in a better fashion.

**Review by MikeAlhayek — APPROVED** [review]

**Comment by MikeAlhayek** [comment]
@microposmp did you try it with a live connection to Azure to ensure it is removing the correct file for the tenant only?

**Comment by microposmp** [comment]
Yes I've tested it against Azure blob storage. Here are a few screenshots from before and after deleting a tenant.

Before:

image

image

image

After:

image

image

File: src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: So as said, this can use `BlobFileStore`. That's the lower-level service which is suitable for any Blob Storage use case.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: yes but we also don't use it when we construct the service. I am okay with using `BlobFileStore` because its better. I am also okay with not using it too since we are already not using it.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/BlobModularTenantEvents.cs:30]: We can't use it under https://github.com/OrchardCMS/OrchardCore/blob/0d57489eca05d1924d6bd0eb56cbda170812c6f5/src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs#L37 because `PersistKeysToAzureBlobStorage` requires a `BlobClient`. But it's not a big deal here either, actually, let's leave it.

**Review by Piedone — APPROVED** [review]

**Pull request merged**
